File: src/ActualLab.Core/Time/RetryDelaySeq.cs
using MessagePack;

namespace ActualLab.Time;

[DataContract, MemoryPackable(GenerateType.VersionTolerant)]
#if NET8_0_OR_GREATER
[MessagePackObject(true, SuppressSourceGeneration = true)]
#else
[MessagePackObject(true)]
#endif
public partial record RetryDelaySeq(
    [property: DataMember, MemoryPackOrder(0)] TimeSpan Min,
    [property: DataMember, MemoryPackOrder(1)] TimeSpan Max,
    [property: DataMember, MemoryPackOrder(2)] double Spread)
{
    public const double DefaultSpread = 0.1;
    public const double DefaultMultiplier = 1.41421356237; // Math.Sqrt(2)

    public static RetryDelaySeq Fixed(double delayInSeconds, double spread = DefaultSpread)
        => Fixed(TimeSpan.FromSeconds(delayInSeconds), spread);
    public static RetryDelaySeq Fixed(TimeSpan delay, double spread = DefaultSpread)
        => new(delay, delay, spread, 1);

    public static RetryDelaySeq Exp(double minInSeconds, double maxInSeconds, double spread = DefaultSpread, double multiplier = DefaultMultiplier)
        => new (TimeSpan.FromSeconds(minInSeconds), TimeSpan.FromSeconds(maxInSeconds), spread, multiplier);
    public static RetryDelaySeq Exp(TimeSpan min, TimeSpan max, double spread = DefaultSpread, double multiplier = DefaultMultiplier)
        => new (min, max, spread, multiplier);

    [DataMember, MemoryPackOrder(3)]
    public double Multiplier { get; init; } = DefaultMultiplier;

    public virtual TimeSpan this[int failureCount] {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => GetDelay(failureCount);
    }

    [JsonConstructor, Newtonsoft.Json.JsonConstructor, MemoryPackConstructor, SerializationConstructor]
    public RetryDelaySeq(
        TimeSpan min, TimeSpan max,
        double spread = DefaultSpread,
        double multiplier = DefaultMultiplier)
        : this(min, max, spread)
        => Multiplier = multiplier;

    public virtual TimeSpan GetDelay(int failureCount)
    {
        if (Min <= TimeSpan.Zero)
            throw new InvalidOperationException(
                $"{nameof(RetryDelaySeq)}.{nameof(Min)} must be greater than zero.");

        if (failureCount <= 0)
            return TimeSpan.Zero;
        if (Multiplier <= 1d) // Fixed, i.e. no exponential component
            return Min.ToRandom(Spread).Next().Positive();

        try {
            var multiplier = Math.Pow(Multiplier, failureCount - 1);
            return TimeSpanExt.Min(Max, Min.MultiplyBy(multiplier)).ToRandom(Spread).Next().Positive();
        }
        catch (OverflowException) {
            return Max;
        }
    }

    public IEnumerable<TimeSpan> Delays(int start = 0, int count = int.MaxValue)
    {
        for (var tryIndex = start; count > 0; tryIndex++, count--)
            yield return this[tryIndex];
    }

    // Conversion

    public override string ToString()
        => Multiplier <= 1d
            ? $"{nameof(RetryDelaySeq)}.{nameof(Fixed)}({Min.ToShortString()} ± {Spread:P0}])"
            : $"{nameof(RetryDelaySeq)}.{nameof(Exp)}({Min.ToShortString()} ... {Max.ToShortString()} ± {Spread:P0}, x {Multiplier:F2}])";
}

File: ts/packages/core/src/retry-delay-seq.ts
// .NET counterpart: ActualLab.Time.RetryDelaySeq

const DefaultSpread = 0.1;
const DefaultMultiplier = 1.41421356237; // Math.sqrt(2)

function applySpread(origin: number, spread: number): number {
  if (spread <= 0) return origin;
  const delta = origin * spread;
  return origin + (Math.random() * 2 - 1) * delta;
}

export class RetryDelaySeq {
  readonly min: number;
  readonly max: number;
  readonly spread: number;
  readonly multiplier: number;

  constructor(min: number, max: number, spread: number = DefaultSpread, multiplier: number = DefaultMultiplier) {
    this.min = min;
    this.max = max;
    this.spread = spread;
    this.multiplier = multiplier;
  }

  static fixed(delayMs: number, spread: number = DefaultSpread): RetryDelaySeq {
    return new RetryDelaySeq(delayMs, delayMs, spread, 1);
  }

  static exp(
    minMs: number,
    maxMs: number,
    spread: number = DefaultSpread,
    multiplier: number = DefaultMultiplier,
  ): RetryDelaySeq {
    return new RetryDelaySeq(minMs, maxMs, spread, multiplier);
  }

  getDelay(failureCount: number): number {
    if (this.min < 0) throw new Error("RetryDelaySeq.min must be non-negative.");
    if (failureCount <= 0) return 0;

    if (this.multiplier <= 1) {
      // Fixed — no exponential component
      return Math.max(0, this.spread <= 0 ? this.min : applySpread(this.min, this.spread));
    }

    const multiplier = Math.pow(this.multiplier, failureCount - 1);
    const raw = Math.min(this.max, this.min * multiplier);
    return Math.max(0, applySpread(raw, this.spread));
  }
}
